photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala

```diff
@@ -119,37 +119,54 @@ object ModelTraining extends Logging {
     // Initialize the broadcasted normalization context
     val broadcastNormalizationContext = trainingData.sparkContext.broadcast(normalizationContext)
 
-    // Choose the generalized linear algorithm
-    val initOptimizationProblem:
-        GeneralizedLinearOptimizationProblem[GeneralizedLinearModel, DiffFunction[LabeledPoint]] = taskType match {
-
-      case LOGISTIC_REGRESSION => LogisticRegressionOptimizationProblem.buildOptimizationProblem(
-        optimizationConfig,
-        treeAggregateDepth,
-        enableOptimizationStateTracker)
-
-      case LINEAR_REGRESSION => LinearRegressionOptimizationProblem.buildOptimizationProblem(
-        optimizationConfig,
-        treeAggregateDepth,
-        enableOptimizationStateTracker)
-
-      case POISSON_REGRESSION => PoissonRegressionOptimizationProblem.buildOptimizationProblem(
-        optimizationConfig,
-        treeAggregateDepth,
-        enableOptimizationStateTracker)
-
-      case SMOOTHED_HINGE_LOSS_LINEAR_SVM => SmoothedHingeLossLinearSVMOptimizationProblem.buildOptimizationProblem(
-        optimizationConfig,
-        treeAggregateDepth,
-        enableOptimizationStateTracker)
+    // Construct the generalized linear optimization problem
+    val (glmConstructor, objectiveFunction) = taskType match {
+      case LOGISTIC_REGRESSION =>
+        val constructor = LogisticRegressionModel.createModel _
+        val objective = DistributedGLMLossFunction.createLossFunction(
```

Comment:
Sounds good. And maybe we also want to document that ```DistributedGLMLossFunction``` (and other loss functions) can actually create objective function?